File: pbmx_ffi/java/io/rmf/pbmx/payloads/BytesPayload.java
package io.rmf.pbmx.payloads;

import io.rmf.pbmx.BlockBuilder;
import java.nio.ByteBuffer;

public final class BytesPayload extends Payload {
    public BytesPayload(byte[] bytes) {
        this.bytes = bytes;
    }

    @Override
    public void addTo(BlockBuilder builder) {
        ByteBuffer buf = ByteBuffer.wrap(this.bytes);
        builder.bytes(buf);
    }

    private final byte[] bytes;
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/payloads/BytesPayload.kt
package io.rmf.pbmx.payloads

import com.sun.jna.Pointer
import com.sun.jna.ptr.LongByReference
import io.rmf.pbmx.ffi.FFI

class BytesPayload(handle: Pointer) : Payload(handle) {
    val bytes: ByteArray

    init {
        val length = LongByReference()
        var r = FFI.pbmx_unwrap_bytes(this.handle, null, length)
        assert(r == 0)

        this.bytes = ByteArray(length.value.toInt())
        r = FFI.pbmx_unwrap_bytes(this.handle, this.bytes, length)
        assert(r != 0)
    }
}
